File: Queue/CircularDeque/circular.py
class MyCircularDeque:

    def __init__(self, k: int):
        self.arr = [0] * (k+1)
        self.front = 0
        self.back = 0

    def insertFront(self, value: int) -> bool:
        if self.isFull(): return False
        self.front = (self.front - 1 + len(self.arr)) % len(self.arr)
        self.arr[self.front] = value
        return True

    def insertLast(self, value: int) -> bool:
        if self.isFull(): return False
        self.arr[self.back] = value
        self.back = (self.back+1)%len(self.arr)
        return True


    def deleteFront(self) -> bool:
        if self.isEmpty(): return False
        self.front = (self.front+1)%len(self.arr)
        return True

    def deleteLast(self) -> bool:
        if self.isEmpty(): return False
        self.back = (self.back - 1 + len(self.arr)) % len(self.arr)
        return True

    def getFront(self) -> int:
        if self.isEmpty(): return -1 
        return self.arr[self.front]

    def getRear(self) -> int:
        if self.isEmpty(): return -1 
        idx = self.back-1
        return self.arr[idx]

    def isEmpty(self) -> bool:
        return self.front == self.back

    def isFull(self) -> bool:
        return (self.back+1)%len(self.arr) == self.front


# Your MyCircularDeque object will be instantiated and called as such:
# obj = MyCircularDeque(k)
# param_1 = obj.insertFront(value)
# param_2 = obj.insertLast(value)
# param_3 = obj.deleteFront()
# param_4 = obj.deleteLast()
# param_5 = obj.getFront()
# param_6 = obj.getRear()
# param_7 = obj.isEmpty()
# param_8 = obj.isFull()

File: Queue/CircularDeque/circular.cpp
#include <iostream>

class Node{
public:
    Node* next;
    Node* prev;
    int val;
    Node(int val){
        this->val = val;
        prev = NULL;
        next = NULL;
    }
};
class MyCircularDeque {
public:
    Node* head;
    Node* tail;
    int size;
    int capacity;
    MyCircularDeque(int k) {
        capacity = k;
        size = 0;
        head = NULL;
        tail = NULL;
    }
    
    bool insertFront(int value) {
        if(size==capacity)return false;
        Node* temp = new Node(value);
        if(size==0){
            head = temp;
            tail = temp;
        }
        else{
            temp->next = head;
            head->prev = temp;
            head = temp;
        }
        size++;
        return true;
    }
    
    bool insertLast(int value) {
        if(size==capacity)return false;
        Node* temp = new Node(value);
        if(size==0){
            head = temp;
            tail = temp;
        }
        else{
            tail->next = temp;
            temp->prev = tail;
            tail = temp;
        }
        size++;
        return true;
    }
    
    bool deleteFront() {
        if(size==0)return false;
        else if(size==1){
            head = NULL;
            tail = NULL;
        }
        else{
            Node* temp = head;
            head = head->next;
            head->prev = NULL;
            delete(temp);
        }
        size--;
        return true;
    }
    
    bool deleteLast() {
        if(size==0)return false;
        else if(size==1){
            head = NULL;
            tail = NULL;
        }
        else{
            Node* temp = tail;
            tail = tail->prev;
            tail->next = NULL;
            delete(temp);
        }
        size--;
        return true;
    }
    
    int getFront() {
        if(size==0)return -1;
        return head->val;
    }
    
    int getRear() {
        if(size==0)return -1;
        return tail->val;
    }
    
    bool isEmpty() {
        if(size)return false;
        return true;
    }
    
    bool isFull() {
        if(size==capacity)return true;
        return false;
    }
};

/**
 * Your MyCircularDeque object will be instantiated and called as such:
 * MyCircularDeque* obj = new MyCircularDeque(k);
 * bool param_1 = obj->insertFront(value);
 * bool param_2 = obj->insertLast(value);
 * bool param_3 = obj->deleteFront();
 * bool param_4 = obj->deleteLast();
 * int param_5 = obj->getFront();
 * int param_6 = obj->getRear();
 * bool param_7 = obj->isEmpty();
 * bool param_8 = obj->isFull();
 */